File: go/game.go
package main

import (

	"bufio"
	"fmt"
	"os"
	"strings"
	"unicode"

)

// gamemode handles test mode functionality
type GameMode struct {

	entries		[][]string	// 6 hands, 5 cards 
	testArray	[]string	// 6 hands as strings
	dupArray	map[string]bool

}

// new game mode
func NewGameMode() *GameMode {

	return &GameMode{

		entries:	make([][]string, 6),
		testArray:	make([]string, 6),
		dupArray:	make(map[string]bool),

	}

}

// read and validate test deck
func (g *GameMode) ReadFile(filename string) error {

	file, err := os.Open(filename)
	if err != nil {

		fmt.Fprintf(os.Stderr, "ERROR! COULD NOT OPEN FILE:  %s\n", filename)
		os.Exit(1)

	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	rowIndex := 0

	// print hands read
	for i := range g.entries {

		for j := range g.entries[i] {

			if j == 4 {

				fmt.Println(g.entries[i][j])

			}else {

				fmt.Print(g.entries[i][j], ",")

			}

		}

	}

	for scanner.Scan() && rowIndex < 6 {

		line := scanner.Text()
		cards := strings.Split(line, ",")

		if len(cards) != 5 {

			fmt.Fprintf(os.Stderr, "*** ERROR - EACH HAND MUST HAVE 5 CARDS ***\n")
			os.Exit(1)

		}

		g.entries[rowIndex] = make([]string, 5)

		for colIndex, card := range cards {

			card = strings.TrimRightFunc(card, unicode.IsSpace)

			// check for duplicates
			if g.dupArray[card] {

				fmt.Fprintf(os.Stderr, "*** ERROR - DUPLICATE CARD FOUND IN DECK ***\n*** DUPLICATE: %s ***\n", card)
				os.Exit(1)

			}

			// check for uppercase suits
			if len(card) == 0 || !isUpperSuit(card) {

				fmt.Fprintf(os.Stderr, "*** ERROR - SUITS MUST BE UPPERCASE ***\nINVALID SUIT: %s\n", card)
				os.Exit(1)

			}

			// check card length 
			if !isValidSpacing(card) {

				fmt.Fprintf(os.Stderr, "*** ERROR - CARDS MUST BE 3 CHARACTERS LONG ***\nINVALID CARD LENGTH: %s\n", card)
				os.Exit(1)

			}

			// validate the face value
			if !isValidFace(card[:len(card)-1]) {

				fmt.Fprintf(os.Stderr, "*** ERROR - INVALID CARD FACE VALUE ***\nINVALID CARD: %s\n", card)
				os.Exit(1)

			}

			// validate the suit
			if !isValidSuit(card[len(card)-1:]) {

				fmt.Fprintf(os.Stderr, "*** ERROR - INVALID CARD SUIT ***\nINVALID CARD: %s\n", card)
				os.Exit(1)

			}

			g.dupArray[card] = true
			g.entries[rowIndex][colIndex] = card

		}

		rowIndex++

	}

	if rowIndex < 6 {

		fmt.Fprintf(os.Stderr, "*** ERROR - FILE MUST CONTAIN 6 HANDS ***\n")
		os.Exit(1)

	}

	if err := scanner.Err(); err != nil {

		fmt.Fprintf(os.Stderr, "*** ERROR READING FILE: %v ***\n", err)
		os.Exit(1)

	}


	g.setTestDeck()
	return nil

}


//////////////// HELPER FUNCTIONS

// check if proper spacing
func isValidSpacing(card string) bool {

	// check for 10
	if strings.HasPrefix(card, "10") {
		return len(card) == 3
	}

	// all other cards
	if len(card) != 3 {
		return false
	}

	return card[0] == ' '

}

// check if suit is uppercase
func isUpperSuit(card string) bool {

	suit := card[len(card)-1:]
	return suit == strings.ToUpper(suit)

}

// check if suit is valid
func isValidFace(face string) bool {

	face = strings.TrimSpace(face)

	validFaces := map[string]bool{

		"A": true, "2": true, "3": true, "4": true, "5": true,
		"6": true, "7": true, "8": true, "9": true, "10": true,
		"J": true, "Q": true, "K": true,

	}

	return validFaces[face]
}

// check if valid suit
func isValidSuit(suit string) bool {

	validSuits := map[string]bool{

		"H": true,
		"D": true,
		"C": true,
		"S": true,

	}

	return validSuits[suit]
}

// prepare test deck
func (g *GameMode) setTestDeck() {

	for i := range g.entries {

		handCards := make([]string, 5)
		for j, card := range g.entries[i] {

			if strings.HasPrefix(card, "10") {
				handCards[j] = card
			} else {
				handCards[j] = card[1:]
			}

		}

		g.testArray[i] = strings.Join(handCards, " ")

	}

	fmt.Println("\n*** Here are the six hands... ")
	for _, hand := range g.testArray {

		fmt.Println(hand)

	}

}

// returns test array of hands
func (g *GameMode) GetTestArray() []string {

	return g.testArray

}

// process test deck
func (g *GameMode) CallDeckClass(d *Deck) {

	d.RankHands(g.testArray)

}

File: rust/src/game.rs
use std::fs::File;                    // For file operations
use std::io::{BufRead, BufReader};    // For reading file contents
use std::process;                     // For program termination
use crate::deck::Deck;                // Import Deck struct

// GameMode struct for handling test deck operations
#[derive(Debug)]
pub struct GameMode {
    pub entries: Vec<Vec<String>>,    // 2D vector to store cards from file (6 hands x 5 cards)
    pub test_array: Vec<String>,      // Array to store formatted hands
    pub dup_array: Vec<String>,       // Array to track duplicate cards
}

impl GameMode {
    // Constructor for GameMode
    pub fn new(_test: bool) -> Self {
        GameMode {
            entries: vec![vec![String::new(); 5]; 6],  // Initialize 6 hands of 5 cards each
            test_array: vec![String::new(); 6],        // Space for 6 formatted hands
            dup_array: Vec::new(),                     // Empty vector for duplicate tracking
        }
    }

    // Reads and validates cards from input file
    pub fn read_file(&mut self, filename: &str) {
        // Open file with error handling
        let file = match File::open(filename) {
            Ok(file) => file,
            Err(_) => {
                eprintln!("ERROR! COULD NOT OPEN FILE: {}", filename);
                process::exit(1);
            }
        };

        let reader = BufReader::new(file);

        // Read up to 6 lines (hands)
        for (row_index, line) in reader.lines().enumerate() {
            if row_index >= 6 {
                break;
            }

            if let Ok(line) = line {
                // Split line into cards
                let tokens: Vec<&str> = line.split(',').collect();
                
                // Process up to 5 cards per hand
                for (col_index, token) in tokens.iter().enumerate() {
                    if col_index >= 5 {
                        break;
                    }

                    let parsed_line = token.to_string();

                    // Check for duplicate cards
                    if self.dup_array.contains(&parsed_line) {
                        eprintln!("*** ERROR - DUPLICATE CARD FOUND IN DECK ***\n*** DUPLICATE: {} ***", parsed_line.trim());
                        process::exit(1);
                    }

                    // Validate suit is uppercase
                    if parsed_line.is_empty() || !parsed_line.chars().last().unwrap_or(' ').is_uppercase() {
                        eprintln!("*** ERROR - SUITS MUST BE UPPERCASE ***\n*** INVALID SUIT: {} ***", parsed_line.trim());
                        process::exit(1);
                    }

                    // Validate card format
                    let is_ten = parsed_line.starts_with("10");
                    // Non-10 cards must be 3 characters (space + face + suit)
                    if !is_ten && parsed_line.len() == 2 {
                        eprintln!("*** ERROR - CARDS MUST BE 3 CHARACTERS LONG ***\n*** INVALID CARD LENGTH: {} ***", parsed_line.trim());
                        process::exit(1);
                    }

                    // Store valid card
                    self.dup_array.push(parsed_line.clone());
                    self.entries[row_index][col_index] = parsed_line;
                }
            }
        }

        // Print all hands from file
        for i in 0..6 {
            for j in 0..5 {
                if j == 4 {
                    println!("{}", self.entries[i][j]);
                } else {
                    print!("{},", self.entries[i][j]);
                }
            }
        }

        self.set_test_deck();
    }

    // Formats hands for poker analysis
    pub fn set_test_deck(&mut self) {
        // Process each hand
        for i in 0..6 {
            // Remove spaces and join cards with spaces
            let line: Vec<String> = self.entries[i]
                .iter()
                .map(|s| s.replace(" ", ""))
                .collect();
            self.test_array[i] = line.join(" ");
        }

        // Print formatted hands
        println!("\n*** Here are the six hands... ");
        for hand in &self.test_array {
            println!("{}", hand);
        }
    }

    // Passes hands to Deck for ranking
    pub fn call_deck_class(&self, deck: &mut Deck) {
        deck.rank_hands(&self.test_array);
    }
}

